iml-agent/src/action_plugins/check_ha.rs

```diff
@@ -71,27 +92,208 @@ impl AgentInfo {
+        .await
         .and_then(|output| {
             // This cannot split between map/map_err because Element does not implement Send
             match Element::from_reader(output.stdout.as_slice()) {
-                Err(err) => future::err(ImlAgentError::XmlError(err)),
-                Ok(elem) => future::ok(
-                    elem.find_all("group")
-                        .flat_map(|g| {
-                            let name = g.get_attr("id").unwrap_or("").to_string();
-                            g.find_all("primitive")
-                                .map(move |p| AgentInfo::create(p, Some(name.clone())))
-                        })
-                        .chain(
-                            elem.find_all("primitive")
-                                .map(|p| AgentInfo::create(p, None)),
-                        )
-                        .collect(),
-                ),
+                Err(err) => Err(ImlAgentError::XmlError(err)),
+                Ok(elem) => Ok(elem
+                    .find_all("group")
+                    .flat_map(|g| {
+                        let name = g.get_attr("id").unwrap_or("").to_string();
+                        g.find_all("primitive")
+                            .map(move |p| AgentInfo::create(p, Some(name.clone())))
+                    })
+                    .chain(
+                        elem.find_all("primitive")
+                            .map(|p| AgentInfo::create(p, None)),
+                    )
+                    .collect()),
             }
         })
-        .from_err()
+}
+
+pub fn get_ha_resource_list(_: ()) -> impl Future<Item = Vec<AgentInfo>, Error = ImlAgentError> {
+    get_ha_resource_list_async(()).boxed().compat()
+}
+/// Output of buctl to just then string:
+/// b "enabled"  ->  enabled
+fn busctl_to_string(output: &str) -> &str {
+    output.split("\"").nth(1).unwrap()
+}
+
+async fn systemd_unit_enabled(name: &str) -> Result<bool, ImlAgentError> {
+    let s = format!("{}.service", name);
+    cmd_output_success(
+        "busctl",
+        &[
+            "--system",
+            "--no-pager",
+            "call",
+            "org.freedesktop.systemd1",
+            "/org/freedesktop/systemd1",
+            "org.freedesktop.systemd1.Manager",
+            "GetUnitFileState",
+            "s",
+            &s,
+        ],
+    )
+    .compat()
+    .await
+    .map(
+        |output| match busctl_to_string(str::from_utf8(&output.stdout).unwrap()) {
+            "enabled" => true,
+            _ => false,
+        },
+    )
+}
+
+/// property: org.freedesktop.systemd1
+/// format!("/org/freedesktop/systemd1/unit/{}_2eservice", name)
+/// UnitFileState   +   ActiveState  -> RunState
+/// disabled            inactive        Stopped
+/// enabled             inactive        Enabled
+/// disabled            active          Started
+/// enabled             active          Setup
+async fn systemd_unit_props(name: &str) -> Result<RunState, ImlAgentError> {
```

Comment:
Instead of unwrapping, use the ? operator to return early:

```rust
async fn systemd_unit_props(name: &str) -> Result<RunState, ImlAgentError> {
    let s = format!("/org/freedesktop/systemd1/unit/{}_2eservice", name);
    let output = cmd_output(
        "busctl",
        &[
            "--system",
            "--no-pager",
            "get-property",
            "org.freedesktop.systemd1",
            &s,
            "org.freedesktop.systemd1.Unit",
            "UnitFileState",
            "ActiveState",
        ],
    )
    .compat()
    .await?;

    let res: Vec<&str> = str::from_utf8(&output.stdout)?
        .lines()
        .map(|o| busctl_to_string(o))
        .collect::<Result<Vec<_>, ImlAgentError>>()?;

    match res[..] {
        ["disabled", "inactive"] => Ok(RunState::Stopped),
        ["enabled", "inactive"] => Ok(RunState::Enabled),
        ["disabled", "active"] => Ok(RunState::Started),
        ["enabled", "active"] => Ok(RunState::Setup),
        _ => Err(ImlAgentError::from(RequiredError(format!(
            "Unknown busctl ({}) output: {:?}",
            s, output.stdout
        )))),
    }
}
```